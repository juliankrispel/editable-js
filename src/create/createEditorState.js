// @flow

import produce from 'immer'
import type { RawContentState, EditorState } from '../types'
import createBlock from './createBlock'

const emptyState = {
  content: [],
  entityMap: {},
  selection: {
    startKey: 'a',
    endKey: 'a',
    startOffset: 0,
    endOffset: 0
  },
  changes: [],
  changeIndex: -1,
  lastCommitted: null
}

const createEditorState = (rawContent: RawContentState): EditorState => {
  const content = rawContent.blocks.map(createBlock)

  const key = content[0].key
  const selection = {
    startKey: key,
    endKey: key,
    startOffset: 0,
    endOffset: 0
  }

  const initialState = produce(emptyState,
    () => ({
      content,
      entityMap: rawContent.entityMap,
      selection,
      changes: [],
      changeIndex: 0,
      lastCommitted: null
    }))

  return initialState
}

export default createEditorState
