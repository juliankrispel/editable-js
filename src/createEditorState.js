// @flow

import genId from './genId'
import type { RawContentState, EditorState } from './types'
import produce from 'immer'

const emptyState = {
  content: [],
  selection: {
    startKey: 'a',
    endKey: 'a',
    startOffset: 0,
    endOffset: 0
  },
  changes: [],
  changeIndex: -1,
  lastCommitted: null
}

const createEditorState = (rawContent: RawContentState): EditorState => {
  const content = rawContent.map(block => ({
    ...block,
    key: genId()
  }))

  const key = content[0].key
  const selection = {
    startKey: key,
    endKey: key,
    startOffset: 0,
    endOffset: 0
  }

  const initialState = produce(emptyState,
    () => ({
      content,
      selection,
      changes: [],
      changeIndex: 0,
      lastCommitted: null
    }))

  return initialState
}

export default createEditorState
