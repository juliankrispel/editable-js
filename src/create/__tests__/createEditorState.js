// @flow
import createEditorState from '../createEditorState'

// const emptyState = {
//   content: [],
//   selection: {
//     startKey: 'a',
//     endKey: 'a',
//     startOffset: 0,
//     endOffset: 0
//   },
//   changes: [],
//   changeIndex: -1,
//   lastCommitted: null
// }

describe('createEditorState', () => {
  test('creates editor state with nested structure', () => {
    expect(createEditorState([{
      text: 'start ',
      key: '1',
      children: [{
        text: 'One',
        key: '2',
        children: [{
          text: 'Two',
          key: '3',
          children: [{
            text: 'Three',
            key: '4'
          }, {
            text: 'end',
            key: '5',
            children: [
              {
                text: 'Last Kid',
                key: '6'
              }
            ]
          }]
        }, {
          text: 'Four',
          key: '7'
        }]
      }, {
        text: 'Five',
        key: '8'
      }]
    }, {
      text: 'Six',
      key: '9'
    }])).toMatchSnapshot()
  })

  test('creates editor state with character data', () => {
    expect(createEditorState([{
      text: 'hello my friend',
      key: '1',
      characterRanges: [{
        start: 0,
        offset: 3,
        styles: ['bold']
      }, {
        start: 5,
        offset: 3,
        styles: ['italic'],
        entities: ['1']
      }]
    }])).toMatchSnapshot()
  })
})
