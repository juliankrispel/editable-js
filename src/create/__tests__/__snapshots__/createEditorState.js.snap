// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createEditorState creates editor state with character data 1`] = `
Object {
  "changeIndex": 0,
  "changes": Array [],
  "content": Array [
    Object {
      "characterData": Array [
        Object {
          "entities": Array [],
          "styles": Array [
            "bold",
          ],
        },
        Object {
          "entities": Array [],
          "styles": Array [
            "bold",
          ],
        },
        Object {
          "entities": Array [],
          "styles": Array [
            "bold",
          ],
        },
        Object {
          "entities": Array [],
          "styles": Array [],
        },
        Object {
          "entities": Array [],
          "styles": Array [],
        },
        Object {
          "entities": Array [
            "1",
          ],
          "styles": Array [
            "italic",
          ],
        },
        Object {
          "entities": Array [
            "1",
          ],
          "styles": Array [
            "italic",
          ],
        },
        Object {
          "entities": Array [
            "1",
          ],
          "styles": Array [
            "italic",
          ],
        },
        Object {
          "entities": Array [],
          "styles": Array [],
        },
        Object {
          "entities": Array [],
          "styles": Array [],
        },
        Object {
          "entities": Array [],
          "styles": Array [],
        },
        Object {
          "entities": Array [],
          "styles": Array [],
        },
        Object {
          "entities": Array [],
          "styles": Array [],
        },
        Object {
          "entities": Array [],
          "styles": Array [],
        },
        Object {
          "entities": Array [],
          "styles": Array [],
        },
      ],
      "children": Array [],
      "entityData": Object {},
      "key": "1",
      "text": "hello my friend",
    },
  ],
  "lastCommitted": null,
  "selection": Object {
    "endKey": "1",
    "endOffset": 0,
    "startKey": "1",
    "startOffset": 0,
  },
}
`;

exports[`createEditorState creates editor state with nested structure 1`] = `
Object {
  "changeIndex": 0,
  "changes": Array [],
  "content": Array [
    Object {
      "characterData": Array [
        Object {
          "entities": Array [],
          "styles": Array [],
        },
        Object {
          "entities": Array [],
          "styles": Array [],
        },
        Object {
          "entities": Array [],
          "styles": Array [],
        },
        Object {
          "entities": Array [],
          "styles": Array [],
        },
        Object {
          "entities": Array [],
          "styles": Array [],
        },
        Object {
          "entities": Array [],
          "styles": Array [],
        },
      ],
      "children": Array [
        Object {
          "characterData": Array [
            Object {
              "entities": Array [],
              "styles": Array [],
            },
            Object {
              "entities": Array [],
              "styles": Array [],
            },
            Object {
              "entities": Array [],
              "styles": Array [],
            },
          ],
          "children": Array [
            Object {
              "characterData": Array [
                Object {
                  "entities": Array [],
                  "styles": Array [],
                },
                Object {
                  "entities": Array [],
                  "styles": Array [],
                },
                Object {
                  "entities": Array [],
                  "styles": Array [],
                },
              ],
              "children": Array [
                Object {
                  "characterData": Array [
                    Object {
                      "entities": Array [],
                      "styles": Array [],
                    },
                    Object {
                      "entities": Array [],
                      "styles": Array [],
                    },
                    Object {
                      "entities": Array [],
                      "styles": Array [],
                    },
                    Object {
                      "entities": Array [],
                      "styles": Array [],
                    },
                    Object {
                      "entities": Array [],
                      "styles": Array [],
                    },
                  ],
                  "children": Array [],
                  "entityData": Object {},
                  "key": "4",
                  "text": "Three",
                },
                Object {
                  "characterData": Array [
                    Object {
                      "entities": Array [],
                      "styles": Array [],
                    },
                    Object {
                      "entities": Array [],
                      "styles": Array [],
                    },
                    Object {
                      "entities": Array [],
                      "styles": Array [],
                    },
                  ],
                  "children": Array [
                    Object {
                      "characterData": Array [
                        Object {
                          "entities": Array [],
                          "styles": Array [],
                        },
                        Object {
                          "entities": Array [],
                          "styles": Array [],
                        },
                        Object {
                          "entities": Array [],
                          "styles": Array [],
                        },
                        Object {
                          "entities": Array [],
                          "styles": Array [],
                        },
                        Object {
                          "entities": Array [],
                          "styles": Array [],
                        },
                        Object {
                          "entities": Array [],
                          "styles": Array [],
                        },
                        Object {
                          "entities": Array [],
                          "styles": Array [],
                        },
                        Object {
                          "entities": Array [],
                          "styles": Array [],
                        },
                      ],
                      "children": Array [],
                      "entityData": Object {},
                      "key": "6",
                      "text": "Last Kid",
                    },
                  ],
                  "entityData": Object {},
                  "key": "5",
                  "text": "end",
                },
              ],
              "entityData": Object {},
              "key": "3",
              "text": "Two",
            },
            Object {
              "characterData": Array [
                Object {
                  "entities": Array [],
                  "styles": Array [],
                },
                Object {
                  "entities": Array [],
                  "styles": Array [],
                },
                Object {
                  "entities": Array [],
                  "styles": Array [],
                },
                Object {
                  "entities": Array [],
                  "styles": Array [],
                },
              ],
              "children": Array [],
              "entityData": Object {},
              "key": "7",
              "text": "Four",
            },
          ],
          "entityData": Object {},
          "key": "2",
          "text": "One",
        },
        Object {
          "characterData": Array [
            Object {
              "entities": Array [],
              "styles": Array [],
            },
            Object {
              "entities": Array [],
              "styles": Array [],
            },
            Object {
              "entities": Array [],
              "styles": Array [],
            },
            Object {
              "entities": Array [],
              "styles": Array [],
            },
          ],
          "children": Array [],
          "entityData": Object {},
          "key": "8",
          "text": "Five",
        },
      ],
      "entityData": Object {},
      "key": "1",
      "text": "start ",
    },
    Object {
      "characterData": Array [
        Object {
          "entities": Array [],
          "styles": Array [],
        },
        Object {
          "entities": Array [],
          "styles": Array [],
        },
        Object {
          "entities": Array [],
          "styles": Array [],
        },
      ],
      "children": Array [],
      "entityData": Object {},
      "key": "9",
      "text": "Six",
    },
  ],
  "lastCommitted": null,
  "selection": Object {
    "endKey": "1",
    "endOffset": 0,
    "startKey": "1",
    "startOffset": 0,
  },
}
`;
